libvoikko/src/voikkogc.1
.TH VOIKKOGC 1 "2008-08-22"
.SH NAME
voikkogc \- test program for Voikko grammar checker
.SH SYNOPSIS
.B voikkogc
.RI [ options ]
.SH DESCRIPTION
.B voikkogc
is a test program for grammar checking functionality in libvoikko, library of Finnish language tools.
It reads sentences or paragraphs from stdin (one per line) and print the results to stdout.
The results are structures containing information about grammar errors found in the input paragraph.
.SH OPTIONS
.TP
.B \-\-tokenize
Instead of looking for grammar errors, split input into tokens. The tokens are prefixed
by type: "W" is a word, "P" is punctuation, "S" is whitespace, "U" is unknown and "E" is
a prefix for error messages.
.TP
.B \-\-split\-sentences
Instead of looking for grammar errors, split input into sentences. The sentences are prefixed
by type: "B" means that end of sentence is a probably correct, "P" means that end of sentence
is a possibly correct (but probably this and the next identified sentence should be joined) and
"E" means that sentence ends at the end of input.
.TP
.BI accept_titles= n
.br
Set the value of the specified boolean option.
.TP
.BI explanation_language= langcode
Print human readable error explanation in the specified language.
.SH BUGS
Human readable error explanations are printed in UTF-8 regardless of current locale
settings.
.SH AUTHOR
voikkogc and this manual page were written by Harri Pitk\[:a]nen (hatapitk@iki.fi).




libvoikko/ChangeLog
2008-08-22 Harri Pitkänen <hatapitk@iki.fi>

	* Modify some grammar error checks to affect 2 instead of
	  1 character so that they become slightly easier to see.
	* Add option explanation_language=... to voikkogc for printing
	  human readable error explanations.
	* Fix whitespace detection in get_char_type.

2008-08-13 Harri Pitkänen <hatapitk@iki.fi>

	* Add check for missing punctuation at the end of sentence.
	* Add option VOIKKO_OPT_ACCEPT_TITLES_IN_GC.

2008-08-10 Harri Pitkänen <hatapitk@iki.fi>

	* Fix character case check after colon.

2008-08-03 Harri Pitkänen <hatapitk@iki.fi>

	* Bug fixing.
	* Add check for duplicated words.

2008-08-02 Harri Pitkänen <hatapitk@iki.fi>

	* Ignore option VOIKKO_OPT_IGNORE_DOT in sentence
	  splitting and grammar checking, but document its effects
	  on string tokenizer.
	* Various small bug fixes.

2008-07-27 Harri Pitkänen <hatapitk@iki.fi>

	* Add support for suggestions from grammar checker.

2008-07-26 Harri Pitkänen <hatapitk@iki.fi>

	* Add character case checks to grammar checker.

2008-07-16 Harri Pitkänen <hatapitk@iki.fi>

	* More punctuation checks to grammar checker.

2008-07-08 Harri Pitkänen <hatapitk@iki.fi>

	* Change grammar checker API to allow more than one
	  grammar error with the same start position.
	* New grammar checker test for space before punctuation.

2008-07-03 Harri Pitkänen <hatapitk@iki.fi>

	* Work on grammar checker.
	* Add whitespace error checks.

2008-07-02 Harri Pitkänen <hatapitk@iki.fi>

	* Work on grammar checker.
	* Increase library version number to 2.0.
	* Update README.

2008-07-01 Harri Pitkänen <hatapitk@iki.fi>

	* Split and update manual pages.
	* Add single paragraph grammar checker cache.

2008-06-29 Harri Pitkänen <hatapitk@iki.fi>

	* Move grammar checking related functionality from
	  voikkohyphenate to a new test tool voikkogc.

2008-06-28 Harri Pitkänen <hatapitk@iki.fi>

	* Add API for grammar checking.

2008-06-08 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.
	* Allow strings longer than LIBVOIKKO_MAX_WORD_CHARS to
	  be converted with voikko_cstrtoucs4.

2008-06-07 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.

2008-06-04 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.
	* Increase version numbers for version 1.8.

2008-05-26 Ville-Pekka Vainio <vpivaini@cs.helsinki.fi>

	* Provide libvoikko.pc for pkg-config.

2008-03-07 Harri Pitkänen <hatapitk@iki.fi>

	* Fix hyphenation of nougat'kaan.

2008-01-29 Harri Pitkänen <hatapitk@iki.fi>

	* Increase cost limit for suggestions from 250 to 350.

2008-01-19 Harri Pitkänen <hatapitk@iki.fi>

	* Iconv on OS X should use UCS-4-INTERNAL too.

2007-12-30 Harri Pitkänen <hatapitk@iki.fi>

	* Some fixes for VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS.
	* Add option VOIKKO_OPT_ACCEPT_MISSING_HYPHENS.

2007-12-29 Harri Pitkänen <hatapitk@iki.fi>

	* Add option VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS.

2007-12-01 Harri Pitkänen <hatapitk@iki.fi>

	* Tokenization functions should not be used without initializing the
	  library.

2007-11-28 Harri Pitkänen <hatapitk@iki.fi>

	* Fix configure test AM_ICONV failing with -Wall on FreeBSD.
	* Tweak URL recognition so that "J.K.Paasikivi" will not be
	  considered as hostname.

2007-11-27 Harri Pitkänen <hatapitk@iki.fi>

	* Add boolean option VOIKKO_OPT_IGNORE_NONWORDS.
	* Do not link voikkospell and voikkohyphenate against libmalaga.

2007-10-30 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-28 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-27 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-21 Harri Pitkänen <hatapitk@iki.fi>

	* Allow the use of environment variable VOIKKO_DICTIONARY_PATH
	  to specify the dictionary path.

2007-10-15 Harri Pitkänen <hatapitk@iki.fi>

	* Add preliminary version of string tokenizer
	  voikko_next_token_ucs4.
	* Increase version numbers for 1.6.

2007-09-08 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement suggestion "." -> ","

2007-08-23 Harri Pitkänen <hatapitk@iki.fi>

	* Change configure.ac to fix build on FreeBSD

2007-08-20 Harri Pitkänen <hatapitk@iki.fi>

	* FreeBSD fix, thanks to Jaakko Heinonen

2007-07-08 Harri Pitkänen <hatapitk@iki.fi>

	* Add option ignore_numbers=n to voikkospell. Change
	  default from 1 to 0 (the library default).
	* Undo 1-to-1 normalisations in spelling suggestions.

2007-06-19 Harri Pitkänen <hatapitk@iki.fi>

	* Add exported functions for freeing the memory
	  allocated for result structures. This is needed
	  in environments where multiple C runtime libraries
	  are used in one process.
	* Increase version numbers for 1.5

2007-06-17 Harri Pitkänen <hatapitk@iki.fi>

	* Generate libvoikko-1.def on Windows to enable
	  dynamic linking of libvoikko-1.dll with code
	  generated using MSVC.
	* Set Finnish locale on Windows to ensure that
	  character case conversions work correctly.

2007-05-07 Harri Pitkänen <hatapitk@iki.fi>

	* Hyphenating "vast'edes" as "vast'e-des" is ugly
	* Hyphenating after ' is wrong
	* Increase version numbers for 1.4.1

2007-04-19 Harri Pitkänen <hatapitk@iki.fi>

	* Check for embedded dots in URL detection
	* Only use URL and number detection when
	  VOIKKO_OPT_NO_UGLY_HYPHENATION is set

2007-04-07 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement _ -> (any character) to OCR suggestions 

2007-03-28 Harri Pitkänen <hatapitk@iki.fi>

	* Update to Automake 1.10
	* Update version information for version 1.4
	* Avoid using rule based hyphenator with certain strings containing
	  numbers and special characters

2007-02-26 Harri Pitkänen <hatapitk@iki.fi>

	* Fix the option VOIKKO_INTERSECT_COMPOUND_LEVEL, which only worked
	  correctly with values 0 and 1. Bug report and suggested fix by
	  Kalle Lampila.
	* Add option intersect_compound_level=n to voikkohyphenate

2007-02-16 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement b <-> h to OCR suggestions


libvoikko/src/voikkogc.c
/* Voikkogc: Testing tool for libvoikko
 * Copyright (C) 2008 Harri Pitkänen <hatapitk@iki.fi>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *********************************************************************************/

#define _GNU_SOURCE

#include "voikko.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <wchar.h>
#include <locale.h>
#include "porting.h"
#ifdef HAVE_NL_LANGINFO
#include <langinfo.h>
#endif // HAVE_NL_LANGINFO

#ifdef HAVE_MBRLEN

void print_tokens(int handle, const char * line) {
	size_t len;
	const char * lineptr;
	mbstate_t mbstate;
	enum voikko_token_type token_type;
	size_t tokenchars;
	len = strlen(line);
	memset(&mbstate, '\0', sizeof(mbstate_t));
	lineptr = line;
	while (len > 0) {
		token_type = voikko_next_token_cstr(handle, lineptr, len, &tokenchars);
		switch (token_type) {
			case TOKEN_WORD:
				printf("W: \"");
				break;
			case TOKEN_PUNCTUATION:
				printf("P: \"");
				break;
			case TOKEN_WHITESPACE:
				printf("S: \"");
				break;
			case TOKEN_UNKNOWN:
				printf("U: \"");
				break;
			case TOKEN_NONE:
				printf("E: unknown token\n");
				return;
		}
		while (tokenchars > 0) {
			size_t charlen = mbrlen(lineptr, len, &mbstate);
			while (charlen > 0) {
				putchar(lineptr[0]);
				lineptr++;
				charlen--;
				len--;
			}
			tokenchars--;
		}
		printf("\"\n");
	}
}


void split_sentences(int handle, const char * line) {
	size_t len;
	const char * lineptr;
	size_t charlen;
	mbstate_t mbstate;
	enum voikko_sentence_type sentence_type;
	size_t sentencechars;
	len = strlen(line);
	memset(&mbstate, '\0', sizeof(mbstate_t));
	lineptr = line;
	while (len > 0) {
		sentence_type = voikko_next_sentence_start_cstr(handle, lineptr, len,
		                &sentencechars);
		switch (sentence_type) {
			case SENTENCE_NONE:
				printf("E: %s\n", lineptr);
				return;
			case SENTENCE_PROBABLE:
				printf("B: ");
				break;
			case SENTENCE_POSSIBLE:
				printf("P: ");
				break;
			case SENTENCE_NO_START:
				// Not returned from this function
				break;
		}
		while (sentencechars > 0) {
			charlen = mbrlen(lineptr, len, &mbstate);
			while (charlen > 0) {
				putchar(lineptr[0]);
				lineptr++;
				charlen--;
				len--;
			}
			sentencechars--;
		}
		printf("\n");
	}
}


void check_grammar(int handle, const char * line, const char * explanation_language) {
	size_t len;
	voikko_grammar_error grammar_error;
	int skiperrors = 0;
	len = strlen(line);
	while (1) {
		grammar_error = voikko_next_grammar_error_cstr(handle, line, len,
		                0, skiperrors);
		if (grammar_error.error_code == 0) {
			printf("-\n");
			return;
		}
		printf("[code=%i, level=0, ", grammar_error.error_code);
		printf("descr=\"\", stpos=%zd, ", grammar_error.startpos);
		printf("len=%zd, suggs={", grammar_error.errorlen);
		if (grammar_error.suggestions) {
			char ** sugg = grammar_error.suggestions;
			while (*sugg) {
				// FIXME: convert from UTF-8
				printf("\"%s\"", *sugg);
				sugg++;
				if (*sugg) printf(",");
			}
			voikko_free_suggest_cstr(grammar_error.suggestions);
		}
		printf("}]");
		if (explanation_language) {
			printf(" (%s)", voikko_error_message_cstr(
				grammar_error.error_code, explanation_language));
		}
		printf("\n");
		skiperrors++;
	}
}

enum operation {TOKENIZE, SPLIT_SENTENCES, CHECK_GRAMMAR};

int main(int argc, char ** argv) {
	size_t bufsize = 10000;
	char * line;
	char * path = 0;
	enum operation op = CHECK_GRAMMAR;
	int handle;
	
	line = malloc(bufsize);
	if (line == 0) {
		printf("E: Out of memory\n");
		return 1;
	}
	
	for (int i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) path = argv[++i];
	}
	const char * voikko_error = (const char *) voikko_init_with_path(&handle, "fi_FI", 0, path);

	if (voikko_error) {
		printf("E: Initialisation of Voikko failed: %s\n", voikko_error);
		free(line);
		return 1;
	}
	
	setlocale(LC_ALL, "");
	char * encoding = nl_langinfo(CODESET);
	voikko_set_string_option(handle, VOIKKO_OPT_ENCODING, encoding);
	
	char * explanation_language = 0;
	
	for (int i = 1; i < argc; i++) {
		if (strncmp(argv[i], "--tokenize", 10) == 0) {
			op = TOKENIZE;
		}
		else if (strncmp(argv[i], "--split-sentences", 17) == 0) {
			op = SPLIT_SENTENCES;
		}
		else if (strcmp(argv[i], "accept_titles=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_TITLES_IN_GC, 1);
		else if (strcmp(argv[i], "accept_titles=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_TITLES_IN_GC, 0);
		else if (strncmp(argv[i], "explanation_language=fi", 23) == 0)
			explanation_language = "fi";
		else if (strncmp(argv[i], "explanation_language=", 21) == 0)
			explanation_language = "en";
	}
	
	while (1) {
		ssize_t chars_read = getline(&line, &bufsize, stdin);
		if (chars_read == -1) break;
		if (chars_read > 0 && line[chars_read - 1] == '\n') {
			line[chars_read - 1] = '\0';
		}
		switch (op) {
			case TOKENIZE:
				print_tokens(handle, line);
				break;
			case SPLIT_SENTENCES:
				split_sentences(handle, line);
				break;
			case CHECK_GRAMMAR:
				check_grammar(handle, line, explanation_language);
		}
	}
	free(line);
	voikko_terminate(handle);
	return 0;
}

#else
int main(int argc, char ** argv) {
	printf("E: This tool is not supported on your operating system.\n");
	return 1;
}
#endif

